File: 2020/javascript/src/day02.js
const { readInput } = require("./utils.js");

function day02(sample = readInput("./data/day02.txt")) {
  let [part1, part2] = [0, 0];

  const lines = sample.split("\n");
  const pattern = /^(\d+)-(\d+)\s+(\w):\s+(\w+)$/;

  lines.forEach((entry) => {
    let [_, low, high, targetChar, pwd] = entry.match(pattern);
    const charCounts = pwd.match(RegExp(`[${targetChar}]`, "g"))?.length;
    low = parseInt(low);
    high = parseInt(high);

    if (charCounts >= low && charCounts <= high) part1++;
    if (xor(pwd[low - 1] === targetChar, pwd[high - 1] === targetChar)) part2++;
  });

  return [part1, part2];
}

function xor(a, b) {
  return a ^ b;
}

module.exports = {
  day02,
};

File: 2020/julia/day02.jl
using BenchmarkTools

input = readlines("input/day02.txt")

# Day 2 ------------------------------------------------------------------

function day02(input::Vector{String})
    part1, part2 = 0, 0
    for line in input
        m = match(r"(\d+)-(\d+)\s+(.):\s+(.+)", line)
        isnothing(m) && break
        low = parse(Int, m.captures[1])
        high = parse(Int, m.captures[2])
        char = m.captures[3][1]
        pw = m.captures[4]
        if low <= count(c -> c == char, pw) <= high
            part1 += 1
        end
        if xor(pw[low] == char, pw[high] == char)
            part2 += 1
        end
    end
    return part1, part2
end

println(day02(input))
@benchmark day02(input)